styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -50,32 +74,115 @@
+            .flatMap(this::pairWithResources)
+            .collect(groupingByConcurrent(
+                ResourceWithInstance::resource,
+                ConcurrentHashMap::new,
+                counting()));
+
+    final List<InstanceState> eligibleInstances =
+        activeStates.parallelStream()
+            .filter(entry -> !timedOutInstances.contains(entry.workflowInstance()))
+            .filter(entry -> shouldExecute(entry.runState()))
+            .collect(toCollection(Lists::newArrayList));
+    Collections.shuffle(eligibleInstances);
+
+    final Consumer<InstanceState> limitAndDequeue = (instance) -> {
+      final Set<String> resourceRefs =
+          workflowResourceReferences.getOrDefault(instance.workflowInstance().workflowId(),
+              emptySet());
+
+      if (resourceRefs.isEmpty()) {
+        sendDequeue(instance);
+      } else {
+        final boolean canRun = resourceRefs.stream()
+            .allMatch(resourceId -> {
+              if (!resources.containsKey(resourceId)) {
+                // todo: send info event, referencing unknown resource
+                return false;
+              }
 
-      else if (shouldExecute(state)) {
-        if (state.data().tries() == 0) {
-          LOG.info("Triggering {}", key.toKey(), state.data().tries());
+              final Resource resource = resources.get(resourceId);
+              final long usage = currentResourceUsage.getOrDefault(resourceId, 0L);
+              return usage < resource.concurrency();
+            });
+
+        if (canRun) {
+          resourceRefs.forEach(id -> currentResourceUsage.computeIfAbsent(id, id_ -> 0L));
+          resourceRefs.forEach(id -> currentResourceUsage.compute(id, (id_, l) -> l + 1));
+          sendDequeue(instance);
         } else {
-          LOG.info("{} executing retry #{}", key.toKey(), state.data().tries());
+          // todo: send info event, could not run because resource limit reached
         }
-        stateManager.receiveIgnoreClosed(Event.dequeue(key));
       }
+    };
+
+    timedOutInstances.forEach(this::sendTimeout);
+    eligibleInstances.forEach(limitAndDequeue);
+  }
+
+  private Stream<ResourceWithInstance> pairWithResources(InstanceState instanceState) {
+    final WorkflowId workflowId = instanceState.workflowInstance().workflowId();
+    final Optional<Workflow> workflowOpt = workflowCache.workflow(workflowId);
+    if (!workflowOpt.isPresent()) {
+      return Stream.empty();
     }
+
+    final List<String> referencedResources = workflowOpt.get().schedule().resources();
+
+    return referencedResources.stream()
+        .map(resource -> ResourceWithInstance.create(resource, instanceState));
+  }
+
+  private Set<String> workflowResources(WorkflowId workflowId) {
+    final Optional<Workflow> workflowOpt = workflowCache.workflow(workflowId);
+    if (!workflowOpt.isPresent()) {
+      return emptySet();
+    }
+
+    return Sets.newHashSet(workflowOpt.get().schedule().resources());
```

Comment:
why is this preferred over `new HashSet<>()`? I can see the use in pre-jdk7 but not since `<>` was ok.